fix: missing album/artist/track in database

M2K3K5 opened the pull request:
fixes #429 by re-fetching missing album, artist, or track data from Spotify when it is not found (or invalid) in the database. This ensures that the relevant data is added to the database and prevents page failures or repeated requests.

M2K3K5 committed:
fix: fetch artist/album/track if invalid in database

Yooooomi commented:
Hello! This is something that is already done at startup. If the behavior at startup does not work, this is the thing to fix. I feel like there is a lot of overhead on basic database requests with your pr.

M2K3K5 committed:
fix: db query getAlbumsWithoutArtist, getHistorySongs handling faulty tracks (album/artist null)

M2K3K5 commented:
@Yooooomi  : I've added a fix for your startup feature to fix the missing artist issue. I also fixed issue #309 by updating the getHistorySongs endpoint to skip problematic tracks that were causing the dashboard to crash due to missing artist/album data.

However, I want to emphasize that fixing the database only on startup is not an ideal long-term solution, since most users don't restart their processes often. For example, if I weren't actively developing the project, I would probably only restart it every few months or when a new version is released.

My new code provides a more dynamic solution-it is designed to handle missing artists during regular operation. From the user's perspective, this process should be seamless, and if an artist isn't found in the database, the system will avoid crashes or errors. The only downside is a slight delay (about a second) during data fetch, but this trade-off significantly improves stability and would only happen if the artist/album is missing in the database (which should not happen in the first place).

Yooooomi commented:
If we want to add a solution, it has to be seamless. Seamless in the duration of the request but also, as you said, at runtime to fix any possible issue. I feel like you're building fixes on top of features and it should not be this way. Features should work in isolation without knowing if there is behavior fixing the database. If we think about a solution, it has to be done on the side of the current project. Thanks a lot, really, for your work. And I hope you can understand my point of view.

M2K3K5 commented:
I see your point. Would you prefer moving the re-fetch logic to a different part of the code, outside of the ArtistModel.find function? Another alternative could be executing fixMissingTrackData as a nightly task.

The issue with only fixing the database at process start is that if a single entry is broken, it will break the entire dashboard and the top sessions for the user. To prevent this, the database re-fetch needs to happen as soon as possible, in my opinion.
